prl-mushr/mushr
# Update devcontainer config to support Foxglove Studio on Codespace

## Status

Blocked on
* https://github.com/prl-mushr/mushr/pull/109
* https://github.com/prl-mushr/mushr_sim/pull/20

## Description
Updates foxglove studio installation & launch scripts to enable using foxglove studio app hosted in a codespace instance. See [this demo video](https://www.youtube.com/embed/OtCwCCvPoxQ) or [this parent discussion](https://github.com/prl-mushr/mushr/discussions/108).

### Notes

2023/06/07 _Things to know about the demo video:_ (1) I prebuilt the codespace instance to save the initial build time for the video. (2) To get the demo working, I made the change shown in https://github.com/prl-mushr/mushr_sim/pull/20 on the codespace instance. (3) The self-hosted foxglove studio web app does not allow loading a layout file (e.g., see [this related thread](https://foxglove.slack.com/archives/C02RFDCG5NK/p1686132391155409) in foxglove slack). There are workarounds (e.g., mentioned in the thread), but I propose to wait for the proper fix as foxglove team is working on fixing this issue (I think--reminder: add a related tick if I find one).

## Steps to Test

Follow these steps to set up Mushr in codespace and access the self-hosted foxglove studio:

1. From `https://github.com/mjyc/mushr`, click the Code drop-down menu.
1. Click on the Codespaces tab.
1. Click Create codespace on `noetic-codespace`.
1. Wait until the initial build is done. It takes a looong time, but you just need to go through it once.
1. Run teleop.launch and foxglove studio server, e.g.,
    1. Select Tasks: Run Task and select Roslaunch teleop.launch to launch the teleop launch file.
    1. Select Tasks: Run Task and select Start Foxglove Studio server to start a self-hosted foxglove studio server.
    1. NOTE: You can run these tasks in terminal as shown in [this part of the demo video](https://youtu.be/OtCwCCvPoxQ?t=16).
1. Open Foxglove Studio and connect
    1. See [this part of the demo video](https://youtu.be/OtCwCCvPoxQ?t=41)

## Acknowledgement

I thank [Ruffin](https://github.com/ruffsl) who solved various tricky problems to get Codespace + Foxglove working. For details, see https://discourse.ros.org/t/repeatable-reproducible-and-now-accessible-ros-development-via-dev-containers/31398
